### PR TITLE
`Development`: Split continuous integration service to lower coupling between server services

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
This PR aims to lower the coupling between server services and to reduce the risk of having circular dependencies for those. It creates smaller services and moves some functionality from services into repositories. While the refactoring should not lead to any issues if all server tests and the e2e tests pass, we should still test that the functionality is still working correctly.

The PR also removes some old workarounds which are not necessary any more such as loading build logs from the CI server. (we store them in the database since quite some time)

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Students

This PR refactors multiple aspects of Artemis that need to be tested:

*Static Code Analysis:*
1. Log in to Artemis as an instructor.
2. Create a  programming exercise for a course exercise with SCA enabled (should work with Java and C on TS1 and with Java and Swift on TS2)
3. Configure the grading of the exercise to grade some of the SCA checks. 
4. Log in a s a student and submit a solution containing bad code (e.g. add an unused import) and make sure the result actually reflects the SCA findings.
5. Check the same thing for exam exercises.

*Triggering Builds:*
1. Create a programming exercise and participate as a student.
2. Find the build plan links on the exercise overview page of the instructor and go to the student's build plan.
3. Manually trigger the build from the UI (e.g. from the page containing all participations for the exercise).
4. Check that the builds are triggered in the CI system.
5. Check that solution and template repository builds are triggered when submitting a change to the tests repository.

*Build Results and Build Logs:*
1. Create a programming exercise and participate as a student.
2. Submit a solution that earns you more than 0 points (e.g. add the bubbleSort implementation for the standard Java exercise: https://github.com/ls1intum/Artemis/blob/develop/src/test/cypress/fixtures/exercise/programming/all_successful/BubbleSort.txt)
3. Make sure that the result received contains a message for the failed test cases as well as logs hinting to what went wrong and correctly distinguishes between failed and passed tests.
6. Submit a failed build (e.g. remove a bracket) and check that build logs are displayed instead of successful and failed tests.
7. Check the same thing for exam exercises.

#### Code Review
- [x] Review 1
- [ ] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test

### Test Coverage
no changes except deletion of unused code

### Screenshots
no changes

### CODE REVIEW

**Review by danielbode — APPROVED** [review]
Tested an exam with programming exercises and regular programming exercises with SCA on ts2. Everything worked as expected. Didn't notice any bugs.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackService.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
+package de.tum.in.www1.artemis.service.programming;
+
+import java.util.*;
```
Comment (1999Markus) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackService.java:3]: Coulde be replaced by 
```
import java.util.HashSet;
import java.util.Set;
import java.util.List;
```

**Review by 1999Markus — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackService.java (hunk 1 above)
Comment (1999Markus) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackService.java:3]: Scimmed through the refactored code, looks good to me.

**Review by laadvo — APPROVED** [review]
Tested on ts1, tried out different programming exercises (including in exam mode). Everything seems to be working as expected.

**Review by frederik-kt — APPROVED** [review]
Tested on TS2. Checked course and exam exercises, also with static code analysis. Results were created correctly, build logs were shown correctly. Also tested triggering of builds from the UI. Template and Solution repository are built automatically after pushing to the tests repository.

**Comment by 4ludwig4** [comment]
Checked on TS2 during the test session today.
The CI build notifications and logs work as they got updated.

What didn't work for me, but which might be related to a configuration or permission problem(?) is the fact that the links to the exercise's repository in the version control system (GitLab in this case) in Artemis and in Jenkins didn't work. Still, Jenkins is somehow able to build something and shows the solution to be correct, and the base to be with a lot of failures.

The above incident happened in the following exercise:
https://artemis-test2.artemis.in.tum.de/course-management/8508/programming-exercises/12153
https://jenkins-test.artemis.in.tum.de/job/TCLPTES/job/TCLPTES-SOLUTION/2/
https://gitlab-test.artemis.in.tum.de/TCLPTES/tclptes-solution.git

**Comment by frederik-kt** [comment]
> Checked on TS2 during the test session today. The CI build notifications and logs work as they got updated.
> 
> What didn't work for me, but which might be related to a configuration or permission problem(?) is the fact that the links to the exercise's repository in the version control system (GitLab in this case) in Artemis and in Jenkins didn't work. Still, Jenkins is somehow able to build something and shows the solution to be correct, and the base to be with a lot of failures.
> 
> The above incident happened in the following exercise: https://artemis-test2.artemis.in.tum.de/course-management/8508/programming-exercises/12153 https://jenkins-test.artemis.in.tum.de/job/TCLPTES/job/TCLPTES-SOLUTION/2/ https://gitlab-test.artemis.in.tum.de/TCLPTES/tclptes-solution.git

I tested this again with `develop` deployed on TS2 and got the same 404 when trying to go to that repository on GitLab. So it is not related to this PR. Still weird though.

**Review by sleiss — APPROVED** [review]
Tested it in the testing session, still works as expected.

**Review by 4ludwig4 — APPROVED** [review]

**Review by mertyldrr — APPROVED** [review]
Tested in the testing session, builds triggered successfully after submitting a change.

**Review by rriyaldhi — APPROVED** [review]
Tested on TS2 during testing session and works as expected.

**Pull request merged**